src/algorithms/LaurentPoly.jl

```diff
@@ -0,0 +1,212 @@
+   c = parent(a) == parent(b)
+   c || !throw || error("incompatible Laurent polynomial rings")
+   return c
+end
+
+
+###############################################################################
+#
+#   Basic manipulation
+#
+###############################################################################
+
+const hashp_seed = UInt === UInt64 ? 0xcdaf0e0b5ade239b : 0x5ade239b
+
+function Base.hash(p::LaurentPolyElem, h::UInt)
+   for i in monomials_degrees(p)
+      c = coeff(p, i)
+      if !iszero(c)
+         h = hash(i, h)
+         h = hash(c, h)
+      end
+   end
+   hash(hashp_seed, h)
+end
+
+# required implementation
+"""
+    monomials_degrees(p::LaurentPolyElem) -> AbstractVector{<:Integer}
+
+> Return a vector containing at least all the degrees of the non-null
+> monomials of `p`.
+"""
+function monomials_degrees end
+
+# return a monomials_degrees vector valid for both polys
+function monomials_degrees(p::LaurentPolyElem, q::LaurentPolyElem)
+   mdp = monomials_degrees(p)
+   mdq = monomials_degrees(q)
+   T = promote_type(eltype(mdp), eltype(mdq))
+   @assert T <: Signed # -one(T) must not wrap around
+   if isempty(mdq)
+      mdp, mdq = mdq, mdp
+   end
+   # if only one is empty, it's now mdp
+   if isempty(mdp) # extrema requires non-empty
+      if isempty(mdq)
+         zero(T):-one(T)
+      else
+         UnitRange{T}(extrema(mdq)...) # mdq not returned for type-stability
+      end
+   else # none is empty
+      minp, maxp = extrema(mdp)
+      minq, maxq = extrema(mdq)
+      min(minp, minq):max(maxp, maxq)
+   end
+end
+
+# like monomials_degrees, but return a "strict" range, whose min/max
+# corresponds to non-zero coeffs, or an empty range for p == 0
+function degrees_range(p::LaurentPolyElem)
+   mds = monomials_degrees(p)
+   T = eltype(mds)
+   minp, maxp = isempty(mds) ? (T(0), T(0)) : extrema(mds)
+   while iszero(coeff(p, minp))
+      minp += 1
+      minp > maxp && return minp:maxp
+   end
+   while iszero(coeff(p, maxp))
+      maxp -= 1
+   end
+   minp:maxp
+end
+
+
+# return the degree of the unique monomial, throw if not a monomial
+function monomial_degree(p::LaurentPolyElem)
+   isnull = true
+   local deg
+   mds = monomials_degrees(p)
+   isempty(mds) && throw(DomainError(p, "not a monomial"))
+   for d in mds
+      if !iszero(coeff(p, d))
+         !isnull && throw(DomainError(p, "not a monomial"))
+         deg = d
+         isnull = false
+      end
+   end
+   deg
+end
+
+function lead(p::LaurentPolyElem)
+   dr = degrees_range(p)
+   isempty(dr) ? zero(base_ring(p)) : coeff(p, last(dr))
+end
+
+function trail(p::LaurentPolyElem)
+   dr = degrees_range(p)
+   isempty(dr) ? zero(base_ring(p)) : coeff(p, first(dr))
+end
+
+gens(R::LaurentPolynomialRing) = [gen(R)]
+
+isgen(p::LaurentPolyElem) = p == gen(parent(p))
+
+# whether p is a (monic) monomial of degree i
+function ismonomial(p::LaurentPolyElem, i::Integer; rec::Bool=true)
```

Comment:
What is the purpose of the `rec` argument?

Reply:
I think the idea was to handle cases where the base ring is itself made of polynomials: e.g. 
```
julia> R, x = PolynomialRing(ZZ, "x")
(Univariate Polynomial Ring in x over Integers, x)

julia> S, y = PolynomialRing(R, "y")
(Univariate Polynomial Ring in y over Univariate Polynomial Ring in x over Integers, y)

julia> p = (x)*y^0
(x)

julia> parent(p) == S
true

julia> ismonomial(p)
true
```
With the "Laurent" version of the above, with `rec == false`, `ismonomial(p, 0, rec=false)` would return `false` because the coefficient of the only term at the `S` layer is `x`, which is not one. With `rec == true`, it would be supposed (if implemented) to return `true` as in the example above.

Reply:
I personally would be in favor to not have the `rec` argument. It is mixing different things.

Reply:
Note that this is an internal method; I gave this name because I found both notions related, but I have renamed it now.